opal/mca/common/ucx/common_ucx.c

```diff
@@ -39,10 +43,13 @@ static void opal_common_ucx_mem_release_cb(void *buf, size_t length,
 
 OPAL_DECLSPEC void opal_common_ucx_mca_var_register(const mca_base_component_t *component)
 {
+    static const char *default_tls = "rc_verbs,ud_verbs,rc_mlx5,dc_mlx5,cuda_ipc,rocm_ipc";
```

Comment:
Do we know that this is a good enough default?  I.e., are there devices that export userspace verbs interfaces (RC or UD) that don't want to use UCX?  Infinipath comes to mind.  I don't know about Omnipath, or others...?

Reply:
that's a good question, but how was it handled by OpenIB BTL?

Reply:
They have their own libraries (psm, psm2) that only support infinipath and omnipath so pml/cm took precedence because their library found their hardware. You could use the openib btl with those networks by simply specifying ```--mca pml ob1```. For UCX you will need to specifically make sure that the priority is low if either network is found (at least lower than pml/cm but probably also lower than pml/ob1).

Reply:
seems we want is:
(1) prio_ucx_with_mlx = prio_cm_with_opa/ipath > prio_ucx_with_verbs_non_mlx > prio_openib_btl
OR:
(2) prio_ucx_with_mlx = prio_cm_with_opa/ipath > prio_openib_btl > prio_ucx_with_verbs_non_mlx
WDYT?

Reply:
I'm not sure what you're asking.  Remember, the openib BTL no longer exists on master (i.e., what will become v5.0.0).

What we want is:

1. CM PML is used for EFA-, GNI-, Infinipath-, Omnipath-, iWARP-, and Portals-based networks
2. UCX PML is used for IB- or RoCE-based networks.
3. OB1 PML is used for everything else.

See this slide: https://docs.google.com/presentation/d/1AFBgQ-O8YEoqgNjXoQvVFK0SNd7iW899JonxHu6UY4Y/edit#slide=id.g8bc052a5e6_1_18

Reply:
The tricky part is that some of (1) networks are also exposing and/or emulating IB capabilities..

We can have a more strict (but ugly) check that the device we use with UCX belongs to a "mlx**" driver for (2)
Another option for (2) is to add UCX API for getting more details about devices, vendors etc- but then it won't work on current/previous UCX versions.
WDYT?

Reply:
@yosefe That approach sounds fine to me. In the long run I wonder if we should revive the idea of a framework to handle the enabling/disabling of network types, APIs, etc. Could bake this logic into that component. btl/uct, for example, also should only be enabling on mellanox hardware. Right now it looks for specific transports but the logic is fragile and could break if UCX changes how transports are named.

Reply:
IMO we need to have a common approach of how to decide which component to select, and we can add the necessary APIs in UCX to support that. Indeed, using transport/device/vendor string name does not feel right for the long term...

Reply:
For 4.1.1, I think I'd propose:

1) If the UCX PML sees a MLNX device (either by device driver permissions or even "there's any PCI device with an MLNX vendorid", UCX PML sets its priority to 52.
2) If the UCX PML sees any other devices, it sets its priority to 19 (below OB1)
3) Libfabric MTL / BTL adds verbs to the default provider blocklist, so that the reverse selection problem doesn't occur.

This is not perfect.  It really sucks.  But it should work well enough for 4.1.x.  And, because we're just goofing with priorities instead of disabling the UCX PML, customers can say `-mca pml ucx` to enable UCX in non-default situations.

Long term, I think we need to get back to some type of centralized network selection infrastructure.  I'm not sure we can do that for 5.0.0, but I do think a centralized vendorid map to set priorities of UCX/OFI/PSM/btls should be possible.  I volunteered at the telecon today to write up a proposal by end of the week.